File: genetic-selectors/src/main/java/genetic/selectors/roulette/RouletteDistancer.java
package genetic.selectors.roulette;

import java.util.TreeMap;
import java.util.function.Function;

import genetic.api.individual.FitnessTested;
import javaslang.Tuple2;
import javaslang.collection.Traversable;


/**
 * @author piotr.larysz
 */
public interface RouletteDistancer<T extends Comparable<T>> {

    <M extends FitnessTested<T>> TreeMap<T, M> distance(Traversable<M> entries);

    interface KeyCreator<T extends Comparable<T>> {

        <M extends FitnessTested<T>> Function<M, Tuple2<T, M>> get();
    }
}

File: genetic-selectors/src/main/java/genetic/selectors/roulette/RouletteDistancer.kt
package genetic.selectors.roulette

import genetic.api.individual.Rated
import java.util.TreeMap

interface RouletteDistancer {

  operator fun <M : Rated> invoke(entries: Sequence<M>): TreeMap<Double, M>

  interface KeyCreator {

    operator fun <M : Rated> invoke(): (M) -> Double
  }
}